### ISSUE TITLE
subpixel smoothing give me not symmetric matrix for epsilon at a fixed location

### ISSUE THREAD

**Issue by dc3505** [NanoComp/meep#1766]
Dear all,

I'm wondering since for isotropic material, I thought the smoothed epsilon matrix will be symmetric. In other words, the epsilon_xy and epsilon_yx component will be the same for a fixed location. 
However, when I run the following code:
```
e2 = np.zeros([101, 101], dtype=np.complex128)
e1 = np.zeros([101, 101], dtype=np.complex128)
x = np.linspace(-1, 1, 101)
y = np.linspace(-1, 1, 101)
for i in range(len(x)):
    for j in range(len(y)):
        v3 = mp.py_v3_to_vec(sim.dimensions, mp.Vector3(x[i], y[j], 0), sim.is_cylindrical)
        e2[i, j] = sim.fields.get_inveps(mp.Ex, mp.Y, v3)
        e1[i, j] = sim.fields.get_inveps(mp.Ey, mp.X, v3)
```
`e1 - e2` will not be 0. May I check why this is the case? Or is my understanding of the paper incorrect?

(Just in case, this is just a follow up question, but I can't reopen my previous issue... sorry for keep creating new ones!)

Thanks!

**Comment by stevengj** [comment]
Yes, the subpixel smoothing produces symmetric ε tensors, but we don't actually store all of those components.

It's tricky, because different components of the ε tensor are stored at different points in the Yee grid, so at a particular point in space we might only have  ε_xy (for example) and the rest of the components of ε (which is symmetric) are not stored for that grid point.

When you call `get_inveps`, it is actually returning the average of the ε_ij components from the nearest grid points.   Since for different ij this is averaging over different grid points, you are effectively getting components of different ε tensors (at different points in space) which is why it is non-symmetric.

There's probably a better way of returning something sensible to the user here…

(Note also that Meep can return both D and E directly, so usually I try to work with the D and E fields rather than ε directly in computations.)

**Comment by dc3505** [comment]
Hi Professor,

Thanks so much for the reply! I'm learning FDTD and trying to implement your paper my own. So, I want to use Meep as a validation for my implementation. Originally I thought I can directly compare the epsilon grid but it seems it's not feasible now. Is it possible for me set the epsilon grid (exact grid instead of points in space), with my own epsilon tensor entries?

Thanks!

**Comment by stevengj** [comment]
> Originally I thought I can directly compare the epsilon grid but it seems it's not feasible now.

If you call `get_inveps` exactly at the Yee grid point where the corresponding ε component is stored, you should get that ε component exactly at from that grid point. 

> And it's impossible to be perfect aligned with the grid at that position. 

It could be that the pixel at that position has zero overlap with the boundary (i.e. the pixel corner just touches the boundary)?  Try slightly changing the radius of the cylinder.

**Comment by dc3505** [comment]
Hi Professor,

If this is the case, then since epsilon_xy should locate at the corners of Yee grid. This is because they are used for interpolation for Dx and Dy which means for a NxN yee grid with domain size 2, if I use np.linspace(-1, 1, N+1), I should get the exact value of epsilon_xy. 
And this is exactly my first experiment where I get a non symmetric matrix where epsilon_xy!= epsilon_yx. May I check where I’m wrong about this? 

Thanks again for your help!!!
